libs/libarchfpga/src/read_xml_arch_file_sg.cpp

```diff
@@ -0,0 +1,139 @@
+#include "read_xml_arch_file_sg.h"
+#include "read_xml_util.h"
+#include "parse_switchblocks.h"
+#include "pugixml_util.hpp"
+#include "arch_error.h"
+#include "switchblock_types.h"
+
+/**
+ * @brief Parses all <sg_link> tags under a <sg_link_list> tag.
+ * 
+ * @param sg_link_list_tag XML node pointing to the <sg_link_list> tag.
+ * @param loc_data Points to the location in the architecture file where the parser is reading. Used for priting error messages.
+ * @return std::vector<t_sg_link> containing information of scatter gather links.
+ */
+static std::vector<t_sg_link> parse_sg_link_tags(pugi::xml_node sg_link_list_tag,
+                                                 const pugiutil::loc_data& loc_data) {
+    std::vector<t_sg_link> sg_link_list;
+    pugiutil::expect_only_children(sg_link_list_tag, {"sg_link"}, loc_data);
+    for (pugi::xml_node node : sg_link_list_tag.children()) {
+        if (strcmp(node.name(), "sg_link") != 0) continue;
+
+        const std::vector<std::string> expected_attributes = {"name", "x_offset", "y_offset", "z_offset", "mux", "seg_type"};
+        pugiutil::expect_only_attributes(node, expected_attributes, loc_data);
+
+        t_sg_link sg_link;
+        sg_link.name = pugiutil::get_attribute(node, "name", loc_data).as_string();
+        sg_link.mux_name = pugiutil::get_attribute(node, "mux", loc_data).as_string();
+        sg_link.seg_type = pugiutil::get_attribute(node, "seg_type", loc_data).as_string();
+
+        int x_offset = pugiutil::get_attribute(node, "x_offset", loc_data, pugiutil::OPTIONAL).as_int();
```

Comment:
Just checking -- does get_attribute.as_int() with an OPTIONAL flag return 0 if it is not in the xml?  I'd comment that if so. And if not, I'm not sure how this works ...
@soheilshahrouz thinks there's another default_value method that might be easier to read.